test/e2e/build_pipeline_test.go

```diff
@@ -0,0 +1,329 @@
+	}
+
+	// Get Configuration's latest ready Revision's Build, and check that the Build was successful.
+	logger.Infof("Revision is ready and serving, checking Build status.")
+	config, err := clients.ServingClient.Configs.Get(names.Config, metav1.GetOptions{})
+	if err != nil {
+		t.Fatalf("Failed to get Configuration after it was seen to be live: %v", err)
+	}
+	rev, err := clients.ServingClient.Revisions.Get(config.Status.LatestReadyRevisionName, metav1.GetOptions{})
+	if err != nil {
+		t.Fatalf("Failed to get latest Revision: %v", err)
+	}
+	names.Revision = rev.Name
+	buildName := rev.Spec.BuildRef.Name
+	logger.Infof("Latest ready Revision is %q", rev.Name)
+	logger.Infof("Revision's Build is %q", buildName)
+	b, err := clients.BuildClient.TestBuilds.Get(buildName, metav1.GetOptions{})
+	if err != nil {
+		t.Fatalf("Failed to get build for latest revision: %v", err)
+	}
+	if cond := b.Status.GetCondition(duckv1alpha1.ConditionSucceeded); cond == nil {
+		t.Fatalf("Condition for build %q was nil", buildName)
+	} else if cond.Status != corev1.ConditionTrue {
+		t.Fatalf("Build %q was not successful", buildName)
+	}
+
+	// Update the Configuration with an environment variable, which should trigger a new revision
+	// to be created, but without creating a new build.
+	if err := updateConfigWithEnvVars(clients, names, []corev1.EnvVar{{
+		Name:  "FOO",
+		Value: "bar",
+	}}); err != nil {
+		t.Fatalf("Failed to update config with environment variables: %v", err)
+	}
+
+	nextRevName, err := getNextRevisionName(clients, names)
+	if err != nil {
+		t.Fatalf("Error waiting for next revision to be created: %v", err)
+	}
+	names.Revision = nextRevName
+
+	nextRev, err := clients.ServingClient.Revisions.Get(nextRevName, metav1.GetOptions{})
+	if err != nil {
+		t.Fatalf("Failed to get latest Revision: %v", err)
+	}
+
+	if diff := cmp.Diff(rev.Spec.BuildRef, nextRev.Spec.BuildRef); diff != "" {
+		t.Fatalf("Unexpected differences in BuildRef: %v", diff)
+	}
+
+	// Update the Configuration's Build with an annotation, which should trigger the creation
+	// of BOTH a Build and a Revision.
+	if err := updateConfigWithBuildAnnotation(clients, names, map[string]string{
+		"testing.knative.dev/foo": "bar",
+	}); err != nil {
+		t.Fatalf("Failed to update config with environment variables: %v", err)
+	}
+
+	nextRevName, err = getNextRevisionName(clients, names)
+	if err != nil {
+		t.Fatalf("Error waiting for next revision to be created: %v", err)
+	}
+	names.Revision = nextRevName
+
+	nextRev, err = clients.ServingClient.Revisions.Get(nextRevName, metav1.GetOptions{})
+	if err != nil {
+		t.Fatalf("Failed to get latest Revision: %v", err)
+	}
+
+	if diff := cmp.Diff(rev.Spec.BuildRef, nextRev.Spec.BuildRef); diff == "" {
+		t.Fatalf("Got matching BuildRef, wanted different: %#v", rev.Spec.BuildRef)
+	}
+}
+
+func TestBuildAndServePipelineRun(t *testing.T) {
```

Comment:
A lot of these two tests seem to be copy/pasted, to check that the configuration is accepted and creates revisions. It might be worth trying to collapse them into a single table-driven test, which could also share with the `Build` test?

Reply:
definitely yeah :angel: